src/Microsoft.OData.Core/UriParser/Binders/KeyBinder.cs

```diff
@@ -41,21 +41,101 @@ internal KeyBinder(MetadataBinder.QueryTokenVisitor keyValueBindMethod)
         /// </summary>
         /// <param name="collectionNode">Already bound collection node.</param>
         /// <param name="namedValues">The named value tokens to bind.</param>
+        /// <param name="model">The model to be used.</param>
         /// <returns>The bound key lookup.</returns>
-        internal QueryNode BindKeyValues(EntityCollectionNode collectionNode, IEnumerable<NamedValue> namedValues)
+        internal QueryNode BindKeyValues(EntityCollectionNode collectionNode, IEnumerable<NamedValue> namedValues, IEdmModel model)
         {
             Debug.Assert(namedValues != null, "namedValues != null");
             Debug.Assert(collectionNode != null, "CollectionNode != null");
+            Debug.Assert(model != null, "model != null");
 
             IEdmEntityTypeReference collectionItemType = collectionNode.EntityItemType;
-            List<KeyPropertyValue> keyPropertyValues = new List<KeyPropertyValue>();
 
             IEdmEntityType collectionItemEntityType = collectionItemType.EntityDefinition();
+            QueryNode keyLookupNode;
 
+            if (TryBindToDeclaredKey(collectionNode, namedValues, model, collectionItemEntityType, out keyLookupNode))
+            {
+                return keyLookupNode;
+            }
+            else if (TryBindToDeclaredAlternateKey(collectionNode, namedValues, model, collectionItemEntityType, out keyLookupNode))

```

Comment:
This does not seem right.  The key binder should not depend on declared key, or alternate key, or any other key in future.